sw/host/opentitanlib/src/image/image.rs

```diff
@@ -265,12 +257,41 @@ impl Image {
         Ok(self.size)
     }
 
-    /// Sets the `signed_reion_end` field to `self.size`.
-    /// Note that this method must be called AFTER all signed extensions are added but BEFORE
-    /// adding any unsigned extensions.
-    pub fn update_signed_region_end(&mut self) -> Result<usize> {
-        self.borrow_manifest_mut()?.signed_region_end = self.size as u32;
-        Ok(self.size)
+    /// Sets the `signed_region_end` field to the end of the last signed extension.
+    ///
+    /// The end of the signed region is computed as the offset of the first extension that is not
+    /// in `signed_ids` or the size of the image if there is no such extension.
+    pub fn update_signed_region(&mut self, signed_ids: &[u32]) -> Result<()> {
+        let image_size = self.size as u32;
+        let mut first_unsigned_ext = 0u32;
+        let manifest = self.borrow_manifest_mut()?;
+
+        let mut ext_table = manifest.extensions.entries;
+        ext_table.sort_by(|a, b| a.offset.cmp(&b.offset));
+        for e in ext_table {
+            // Ignore any extensions that haven't been added to the image.
+            if e.offset == 0 {
+                continue;
+            }
+            if signed_ids.contains(&e.identifier) {
+                // Since extensions are sorted by offset, if we've already seen an unsigned
+                // extension then the signed region is discontinuous.
+                ensure!(
+                    first_unsigned_ext == 0,
+                    ImageError::MisplacedSignedExtension(e.identifier)
+                );
```

Comment:
nit: since we should never encounter this case shouldn't this be an assertion?

Reply:
You can hit this by providing a raw extension entry with a fixed offset in `manifest_ext.hjson`.